### DISCUSSION TITLE
Format and Operators precedences

### DISCUSSION THREAD

**Discussion by mmatera** [Mathics3/mathics-core#810] [Show and tell]
From a discussion we have in #809, and #802, it came up that (among many other things) the use of the operators precedences in Mathics does not fulfills the WL standard (AKA the WMA behavior).  Recently, I added the `PrecedenceForm` symbol to the builtin list, and started to provide some documentation. However, after the discussion in #809, I check more deeply the current behaviour in Mathics, and I discover that it does not matches with WMA. To bring it closer, I started the following branch:

https://github.com/Mathics3/mathics-core/tree/infix_precedence_fix

which still does not pass the CI. Then, came up the following issues:

* In WMA,  nested expressions with the same precedence are parenthesized. For example,
```
Imply[a, Imply[b,c]]
```
is formatted as `a => (b=> c)` in WMA, but `a=>b=>c` in Mathics.

* In WMA, `Infix[Operands_List, operators_List, ... ]`  does now works like in Mathics: 
```
Infix[{a,b,c},{"+","-"}]
```
produces in WMA  ```a {+,-} b {+,-} c```  instead of `a + b - c` (Mathics behavior)

* In WMA, `Plus` and `Times` do not seem to have  *format* rules, but just `MakeBoxes` rules. This is related to another refactor that is in progress now, but maybe it is useful to be aware of this if we want to improve the compatibility using the current (internal) API.

**Comment by axkr** [comment]
The operators often need a "grouping"  ("left associative", "right associative").

See:
- https://reference.wolfram.com/language/tutorial/OperatorInputForms.html
> The third columns in the tables show how multiple occurrences of a single input form, or of several input forms with the same precedence, are grouped. For example, `a/b/c` is grouped as `(a/b)/c` ("left associative"), while `a^b^c` is grouped as `a^(b^c)` ("right associative"). 
> No grouping is needed in an expression like `a+b+c`, since  `Plus` is fully associative, as represented by the attribute `Flat`.

**Reply by rocky** [reply]
And whether they are binary, unary, or ternary. Robert's tables and discussion go into this.

And apparently it can get more complicated than this because in a ternary operator, there may several symbols involved.

**Reply by mmatera** [reply]
Indeed: left and right associativity is something that we need to re-implement, as well as ternary operators. In any case, I think that as a first step we can fix the goal on passing the full CI again by rewritting 

* MakeBoxes[Infix[...]]  and make_boxes_infix
* Plus.format, Times.format and Power.format  (and maybe Minus and Divide format rules)

Notice that https://github.com/Mathics3/mathics-core/tree/infix_precedence_fix  contains most of the work on that first step. 
I would like to continue with the arithmetic refactor (#809 was in that direction), so I do not plan to work on this right now. However, I think is a kind of "simple" line of work that someone else could take. In that case, I would be happy to help.

**Reply by rocky** [reply]
> so I do not plan to work on this right now.

This would be good for you to not work on this. I would like to think about this, but things are a bit busy for me so I can't.
